test/Microsoft.AspNet.Razor.Runtime.Test/BufferEntryCollectionTest.cs

```diff
@@ -0,0 +1,160 @@
+            Assert.Same(inner.BufferEntries, collection.BufferEntries[2]);
+        }
+
+        [Fact]
+        public void AddChar_ThrowsIfIndexIsOutOfBounds()
+        {
+            // Arrange
+            var collection = new BufferEntryCollection();
+
+            // Act and Assert
+            var ex = Assert.Throws<ArgumentOutOfRangeException>(
+                () => collection.Add(new[] { 'a', 'b', 'c' }, -1, 2));
+            Assert.Equal("index", ex.ParamName);
+        }
+
+        [Fact]
+        public void AddChar_ThrowsIfCountWouldCauseOutOfBoundReads()
+        {
+            // Arrange
+            var collection = new BufferEntryCollection();
+
+            // Act and Assert
+            var ex = Assert.Throws<ArgumentOutOfRangeException>(
+                () => collection.Add(new[] { 'a', 'b', 'c' }, 1, 3));
+            Assert.Equal("count", ex.ParamName);
+        }
+
+        public static IEnumerable<object[]> AddWithChar_RepresentsStringsAsChunkedEntriesData
+        {
+            get
+            {
+                var charArray1 = new[] { 'a' };
+                var expected1 = new[] { "a" };
+                yield return new object[] { charArray1, 0, 1, expected1 };
+
+                var charArray2 = Enumerable.Repeat('a', 10).ToArray();
+                var expected2 = new[] { new string(charArray2) };
+                yield return new object[] { charArray2, 0, 10, expected2 };
+
+                var charArray3 = Enumerable.Repeat('b', 1024).ToArray();
+                var expected3 = new[] { new string('b', 1023) };
+                yield return new object[] { charArray3, 1, 1023, expected3 };
+
+                var charArray4 = Enumerable.Repeat('c', 1027).ToArray();
+                var expected4 = new[] { new string('c', 1024), "cc" };
+                yield return new object[] { charArray4, 1, 1026, expected4 };
+
+                var charArray5 = Enumerable.Repeat('d', 4099).ToArray();
+                var expected5 = new[] { new string('d', 1024), new string('d', 1024), new string('d', 1024), new string('d', 1024), "d" };
+                yield return new object[] { charArray5, 2, 4097, expected5 };
+
+                var charArray6 = Enumerable.Repeat('e', 1025).ToArray();
+                var expected6 = new[] { "ee" };
+                yield return new object[] { charArray6, 1023, 2, expected6 };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(AddWithChar_RepresentsStringsAsChunkedEntriesData))]
+        public void AddWithChar_RepresentsStringsAsChunkedEntries(char[] value, int index, int count, IList<object> expected)
+        {
+            // Arrange
+            var collection = new BufferEntryCollection();
+
+            // Act
+            collection.Add(value, index, count);
+
+            // Assert
+            Assert.Equal(expected, collection.BufferEntries);
+        }
+
+        public static IEnumerable<object[]> Enumerator_TraversesThroughBufferData
```

Comment:
Data shouldn't be performing test actions inside of it, it should just be data. Move all the data bits into their own methods for validation or re-cosntruct the test to do the action bits for you.